src/io/outputs/include/antares/io/outputs/storage.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+#include <stdexcept>
+#include <type_traits>
+
+#include "antares/io/outputs/columns.h"
+
+template<typename T>
+struct is_optional: std::false_type
+{
+};
+
+template<typename U>
+struct is_optional<std::optional<U>>: std::true_type
+{
+};
+
+template<typename T>
+inline constexpr bool is_optional_v = is_optional<T>::value;
+
+// Primary template: applies to any type that's NOT an optional
+template<typename T>
+struct unwrap_optional
+{
+    using type = T; // just return the type unchanged
+};
+
+// Specialization for std::optional<U>
+template<typename U>
+struct unwrap_optional<std::optional<U>>
+{
+    using type = U; // remove the optional and return the inner type
+};
+
+template<typename T>
+using unwrap_optional_t = typename unwrap_optional<T>::type;
+
+class ColumnBasedStorage
+{
+public:
+    void addStringColumn(const std::string& name)
+    {
+        addColumn<StringColumn>(name);
+    }
+
+    void addIntColumn(const std::string& name)
+    {
+        addColumn<IntColumn>(name);
+    }
+
+    void addDoubleColumn(const std::string& name)
+    {
+        addColumn<DoubleColumn>(name);
+    }
+
+    template<typename T>
+    void addOptionalColumn(const std::string& name)
+    {
+        addColumn<OptionalColumn<T>>(name);
+    }
+
+    template<typename T>
+    void addValue(const std::string& column_name, const T& value)
+    {
+        if constexpr (std::is_same_v<T, std::string>)
+        {
+            getColumn<StringColumn>(column_name).add(value);
+        }
+        else if constexpr (std::is_integral_v<T>)
+        {
+            getColumn<IntColumn>(column_name).add(static_cast<int>(value));
+        }
+        else if constexpr (std::is_floating_point_v<T>)
+        {
+            getColumn<DoubleColumn>(column_name).add(static_cast<double>(value));
+        }
+        else if constexpr (is_optional_v<T>)
+        {
+            using Inner = unwrap_optional_t<T>;
+            getColumn<OptionalColumn<Inner>>(column_name).add(value);
+        }
+        else
+        {
+            throw std::runtime_error("Unsupported type"); // TODO
+        }
```

Comment:
this looks a bit too complex, probably too many cases to handle

Reply:
Also, `unwrap_optional_t` is useless, you can straight use `std::optional<T>::value_type` and
```
template <typename T>
struct is_std_optional : std::false_type {};

template <typename T>
struct is_std_optional<std::optional<T>> : std::true_type {};
```
to determine if T is an `std::optional`

Reply:
T is optional, so what is optional<T>? optional<optional<U>>

Reply:
unwrap structs allows to extract the type behind optional.

Reply:
4 cases is that complex for you?

Reply:
In my opinion you might as well remove the template & overload `addValue`
```cpp
void addValue(const std::string& column_name, int value);
// same for double, std::string, std::optional<T> (templated)
```
So you can get rid of `if constexpr (...)`